docs/docs/modelling-the-domain.md

```diff
@@ -4,12 +4,114 @@ title: Modelling your Domain in Sequent
+  commands.rb              # All post command go here
+  events.rb                # All post events go here
+  post_command_handler.rb  # Subscribes to post commands and dispatches post events
+  post.rb                  # The aggregate root
+post.rb                    # Requires the entire aggregate root
+```
+
+### Adding a command
+
+Changes to state start by executing a command. Commands are quite simple classes containing some attributes and attribute validations. Looking at `lib/post/commands.rb` we have one command:
+
+```ruby
+class AddPost < Sequent::Command
+  attrs author: String, title: String, content: String
+  validates_presence_of :author, :title, :content
+end
+```
+
+Let's add a `PublishPost` command to take our `Post` from draft to published. The command will look like this:
+
+```ruby
+class PublishPost < Sequent::Command
+  attrs publication_date: DateTime
+  validates_presence_of :publication_date
+end
+```
+
+We only need the `publication_date` attribute. Commands always target an aggregate, so we already know what to change by its `aggregate_id`. We could set a `publish` flag, but the event already communicates this intent.
+
+_Learn all about commands in the [Command](/docs/concepts/command.html) concept guide._
+{: .notice}
+
+### Handling our new command
+
+The `PostCommandHandler` subscribes to `Post` commands and calls the domain (i.e. the `Post` aggregate root). We can see this happening for `AddPost`:
+
+```ruby
+class PostCommandHandler < Sequent::CommandHandler
+  on AddPost do |command|
+    repository.add_aggregate Post.new(command)
+  end
+end
+```
+
+Because we are adding a new aggregate `add_aggregate` is called with `Post.new(command)` as its argument. The actual business logic of what `Post` looks like is contained in the aggregate root.
+
+We can add our own `on` block below the `AddPost` one:
+
+```ruby
+on PublishPost do |command|
+  do_with_aggregate(command, Post) do |post|
+    post.publish(command.publication_date)
+  end
+end
+```
+
+Sequent retrieves the post for us and we call the (to be defined) `publish` method on the returned `Post` instance.
+
+_Learn all about command handlers in the [CommandHandler](/docs/concepts/command-handler.html) concept guide._
+{: .notice}
 
 ### The Aggregate Root
 
-In `lib/post/post.rb` we find the aggregate root.
+In `lib/post/post.rb` we find the aggregate root. This class encapsulates your business logic. Events are applied to instances of `Post` to give it its current state. We can see here that create a new `Post` will apply multiple events. Besides `PostAdded` we're also applying events to change the author, title and content. You might be tempted to group all those fields in one event, which can be a good idea if those fields always change together. We're using multiple events to emphasise that a single command does not alway correlate to a single event.
+
+```ruby
+class Post < Sequent::AggregateRoot
+  def initialize(command)
```

Comment:
What happens here when an aggregate is reconstructed from the event stream? Does the command somehow get passed in again automatically? Bit unclear about how this method is used later in the lifecycle when `PostAdded` has already happened and we are now handling `PostUpdated` or similar.

Reply:
When reconstructing from the event stream we call rubies `allocate()` method. This "constructs" an object without calling the initialiser.

See [AggregateRoot.load_from_history](https://github.com/zilverline/sequent/blob/bf75eb86c02a3dfc2884872b55760eefaae2a7e0/lib/sequent/core/aggregate_root.rb#L46)

Reply:
Ah, I see. Never used this method before!